python/ray/experimental/signal.py

```diff
@@ -108,23 +123,27 @@ def receive(sources, timeout=10**12):
     answers = ray.worker.global_worker.redis_client.execute_command(query)
     if not answers:
         return []
-    # There will be one answer per source. If there is no signal for a given
-    # source, redis will return an empty list for that source.
-    assert len(answers) == len(sources)
+    # There will be at most one answer per source. If there is no signal for
+    # a given source, redis will provide no answer for that source.
+    # Map the ID of each source s in sources to s itself.
+    source_id_to_idx = dict()
+    for s in sources:
+        source_id_to_idx[_get_task_id(s).hex()] = s
 
     results = []
-    # Decoding is a little bit involved. Iterate through all the sources:
+    # Decoding is a little bit involved. Iterate through all the answers:
     for i, answer in enumerate(answers):
-        # Make sure the answer corresponds to the source
-        assert ray.utils.decode(answer[0]) == _get_task_id(sources[i]).hex()
-        # The list of results for that source is stored in answer[1]
+        # Make sure the answer corresponds to a source, s, in sources.
+        assert ray.utils.decode(answer[0]) in source_id_to_idx
```

Comment:
this assert carries the same information (and would raise almost the same error) as the next line, so I'd remove it